xapian-core/api/registry.cc

```diff
@@ -175,37 +171,36 @@ Registry::Internal::add_defaults()
-    wtschemes_short["coord"] = weighting_scheme_short;
-    weighting_scheme_short = new Xapian::DLHWeight;
-    wtschemes_short["dlh"] = weighting_scheme_short;
-    weighting_scheme_short = new Xapian::DPHWeight;
-    wtschemes_short["dph"] = weighting_scheme_short;
-    weighting_scheme_short = new Xapian::InL2Weight;
-    wtschemes_short["inl2"] = weighting_scheme_short;
-    weighting_scheme_short = new Xapian::IfB2Weight;
-    wtschemes_short["ifb2"] = weighting_scheme_short;
-    weighting_scheme_short = new Xapian::IneB2Weight;
-    wtschemes_short["ineb2"] = weighting_scheme_short;
-    weighting_scheme_short = new Xapian::LMWeight;
-    wtschemes_short["lm"] = weighting_scheme_short;
-    weighting_scheme_short = new Xapian::PL2Weight;
-    wtschemes_short["pl2"] = weighting_scheme_short;
-    weighting_scheme_short = new Xapian::PL2PlusWeight;
-    wtschemes_short["pl2plus"] = weighting_scheme_short;
-    weighting_scheme_short = new Xapian::TfIdfWeight;
-    wtschemes_short["tfidf"] = weighting_scheme_short;
-    weighting_scheme_short = new Xapian::TradWeight;
-    wtschemes_short["trad"] = weighting_scheme_short;
+    weighting_scheme = new Xapian::BB2Weight;
+    wtschemes_short[weighting_scheme->short_name()] = weighting_scheme;
+    weighting_scheme = new Xapian::BM25Weight;
+    wtschemes_short[weighting_scheme->short_name()] = weighting_scheme;
+    weighting_scheme = new Xapian::BM25PlusWeight;
+    wtschemes_short[weighting_scheme->short_name()] = weighting_scheme;
+    weighting_scheme = new Xapian::BoolWeight;
+    wtschemes_short[weighting_scheme->short_name()] = weighting_scheme;
+    weighting_scheme = new Xapian::CoordWeight;
+    wtschemes_short[weighting_scheme->short_name()] = weighting_scheme;
+    weighting_scheme = new Xapian::TradWeight;
+    wtschemes_short[weighting_scheme->short_name()] = weighting_scheme;
+    weighting_scheme = new Xapian::TfIdfWeight;
+    wtschemes_short[weighting_scheme->short_name()] = weighting_scheme;
+    weighting_scheme = new Xapian::InL2Weight;
+    wtschemes_short[weighting_scheme->short_name()] = weighting_scheme;
+    weighting_scheme = new Xapian::IfB2Weight;
+    wtschemes_short[weighting_scheme->short_name()] = weighting_scheme;
+    weighting_scheme = new Xapian::IneB2Weight;
+    wtschemes_short[weighting_scheme->short_name()] = weighting_scheme;
+    weighting_scheme = new Xapian::DLHWeight;
+    wtschemes_short[weighting_scheme->short_name()] = weighting_scheme;
+    weighting_scheme = new Xapian::PL2PlusWeight;
+    wtschemes_short[weighting_scheme->short_name()] = weighting_scheme;
+    weighting_scheme = new Xapian::PL2Weight;
+    wtschemes_short[weighting_scheme->short_name()] = weighting_scheme;
+    weighting_scheme = new Xapian::DPHWeight;
+    wtschemes_short[weighting_scheme->short_name()] = weighting_scheme;
+    weighting_scheme = new Xapian::LMWeight;
+    wtschemes_short[weighting_scheme->short_name()] = weighting_scheme;
```

Comment:
May be I'm missing something but the object that we are referring to here is really a pointer to object and if we do not create a second one we'll have to store the same object pointer in the second map as well and this leads to a lot of errors reported by valgrind for many tests all of which originate from line 239 in this file which is trying to delete the pointer already deleted in line 233.

```cpp
Xapian::Weight * weighting_scheme;
weighting_scheme = new Xapian::BB2Weight;
wtschemes[weighting_scheme->name()] = weighting_scheme;
wtschemes_short[weighting_scheme->short_name()] = weighting_scheme;
```